Distributed codegen drops program order between comm dispatches, so a wait is scheduled before its own rank's send (deterministic P>1 deadlock)

**Component:** Codegen
**Git Commit ID:** `71020585278b68f56c72c40d5570f07dbb20bc8b` (measured here; the three relevant files are byte-identical on current `main` `b10cae3d`, so it applies there too)
**NPU Kind:** Ascend 910B
**Host Platform:** Linux (aarch64)

### Description

A chip dispatch is lowered to one `orch.submit_next_level` call, and the runtime derives that task's dependencies **only from the tensor tags in its `TaskArgs`**. A comm window is passed as an opaque `Tensor.make(..., child_memory=True)`, so the peer-side effect of `pld.tile.remote_store` / `pld.system.notify` / `pld.system.wait` is invisible to that dependency analysis.

Two dispatches on the same rank that interact *only* through a comm window therefore have no edge between them, and **the program order written in `host_orch` is not preserved anywhere**. The scheduler routes a task to its per-worker FIFO as soon as it is READY, and a dispatch whose only job is to wait usually has **no producer at all** — so it is routed *ahead of that rank's own still-pending send*. Since a worker runs one task at a time, the spin-wait owns the core and the send never runs.

The result is a deterministic deadlock in any program where every rank waits and at least one rank's wait lives in a different dispatch from its send.

### Steps to Reproduce

1. Save the `repro.py` below (two ranks, trivial `t + t` compute, one comm ring in each direction; only `pypto` and `torch` are imported).
2. `LD_PRELOAD=<cann>/aarch64-linux/lib64/libhccl.so python3 repro.py 0,1`
3. The run hangs, then raises `finalize_native_run failed with code -100`.

Per rank the program is:

```
dispatch 1  compute
dispatch 2  SEND    (remote_store + notify to the peer)
dispatch 3  compute
dispatch 4  WAIT    (pld.system.wait on the peer's notify)
dispatch 5  compute
```

**There is no cyclic dependency.** Both ranks send at dispatch 2 with nothing in front of the send, and only then wait at dispatch 4, so both notifies are issued long before either rank waits. No ordering a user can reason about makes this program deadlock.

<details>
<summary><code>repro.py</code></summary>

```python
#!/usr/bin/env python3
"""Minimal reproducer: a rank's `pld.system.wait` dispatch is scheduled BEFORE that same
rank's `remote_store`+`notify` dispatch, so two ranks deadlock.

Two ranks, trivial `t+t` compute, one comm ring in each direction. Per rank the program is:

    dispatch 1  compute
    dispatch 2  SEND    (remote_store + notify to the peer)
    dispatch 3  compute
    dispatch 4  WAIT    (pld.system.wait on the peer's notify)
    dispatch 5  compute

Note there is NO cyclic dependency: **both ranks send at dispatch 2 with nothing in front of
the send, and only then wait at dispatch 4.** Both notifies are therefore issued long before
either rank waits, and this program cannot deadlock for any ordering reason.

It deadlocks anyway, on both ranks, with

    PTO2_ERROR_SCHEDULER_TIMEOUT (finalize_native_run failed with code -100)
    sub_class=S1:running-stalled completed=0/1 running=1 ready=0 waiting=0 orch_done=1

and the error line reports `dispatch_id=2` -- the SECOND dispatch on the rank -- while a
healthy run reports `dispatch_id=5`.

Run:  LD_PRELOAD=<cann>/lib64/libhccl.so python3 repro.py 0,1
Expect: hangs, then raises with code -100. A fixed build completes in ~2s and prints
`fwd_ok=True rev_ok=True`.
"""

from __future__ import annotations

import sys
import time

import torch

import pypto.language as pl
import pypto.language.distributed as pld
from pypto import ir
from pypto.ir.distributed_compiled_program import DistributedConfig

D, P, S = 16, 2, 8


def build():
    @pl.program
    class TwoRing:
        @pl.function(type=pl.FunctionType.InCore)
        def compute(
            self,
            X: pl.Tensor[[D, D], pl.FP32],
            Out: pl.Out[pl.Tensor[[D, D], pl.FP32]],
        ) -> pl.Tensor[[D, D], pl.FP32]:
            t = pl.load(X, [0, 0], [D, D])
            return pl.store(pl.add(t, t), [0, 0], Out)

        @pl.function(type=pl.FunctionType.InCore)
        def send(
            self,
            X: pl.Tensor[[D, D], pl.FP32],
            Echo: pl.Out[pl.Tensor[[D, D], pl.FP32]],
            dst: pld.DistributedTensor[[D, D], pl.FP32],
            signal: pld.DistributedTensor[[S, 1], pl.INT32],
            peer: pl.Scalar[pl.INT32],
        ) -> pl.Tensor[[D, D], pl.FP32]:
            t = pl.load(X, [0, 0], [D, D])
            Echo = pl.store(t, [0, 0], Echo)
            pld.tile.remote_store(t, target=dst, peer=peer, offsets=[0, 0])
            pld.system.notify(target=signal, peer=peer, offsets=[0, 0], value=1,
                              op=pld.NotifyOp.AtomicAdd)
            return Echo

        @pl.function(type=pl.FunctionType.InCore)
        def recv(
            self,
            OutD: pl.Out[pl.Tensor[[D, D], pl.FP32]],
            dst: pld.DistributedTensor[[D, D], pl.FP32],
            signal: pld.DistributedTensor[[S, 1], pl.INT32],
        ) -> pl.Tensor[[D, D], pl.FP32]:
            pld.system.wait(signal=signal, offsets=[0, 0], expected=1, cmp=pld.WaitCmp.Ge)
            return pl.store(pl.load(dst, [0, 0], [D, D]), [0, 0], OutD)

        @pl.function(type=pl.FunctionType.Orchestration)
        def c_compute(
            self,
            X: pl.Tensor[[D, D], pl.FP32],
            Out: pl.Out[pl.Tensor[[D, D], pl.FP32]],
        ) -> pl.Tensor[[D, D], pl.FP32]:
            return self.compute(X, Out)

        @pl.function(type=pl.FunctionType.Orchestration)
        def c_send(
            self,
            X: pl.Tensor[[D, D], pl.FP32],
            Echo: pl.Out[pl.Tensor[[D, D], pl.FP32]],
            dst: pld.DistributedTensor[[D, D], pl.FP32],
            signal: pld.DistributedTensor[[S, 1], pl.INT32],
            peer: pl.Scalar[pl.INT32],
        ) -> pl.Tensor[[D, D], pl.FP32]:
            return self.send(X, Echo, dst, signal, peer)

        @pl.function(type=pl.FunctionType.Orchestration)
        def c_recv(
            self,
            OutD: pl.Out[pl.Tensor[[D, D], pl.FP32]],
            dst: pld.DistributedTensor[[D, D], pl.FP32],
            signal: pld.DistributedTensor[[S, 1], pl.INT32],
        ) -> pl.Tensor[[D, D], pl.FP32]:
            return self.recv(OutD, dst, signal)

        @pl.function(level=pl.Level.HOST, role=pl.Role.Orchestrator)
        def host_orch(
            self,
            X: pl.Tensor[[P, D, D], pl.FP32],
            Echo: pl.Out[pl.Tensor[[P, D, D], pl.FP32]],
            Rf: pl.Out[pl.Tensor[[P, D, D], pl.FP32]],
            Rb: pl.Out[pl.Tensor[[P, D, D], pl.FP32]],
        ):
            fdst_buf = pld.alloc_window_buffer(D * D * 4)
            fsig_buf = pld.alloc_window_buffer(S * 4)
            bdst_buf = pld.alloc_window_buffer(D * D * 4)
            bsig_buf = pld.alloc_window_buffer(S * 4)

            T0 = pl.create_tensor([P, D, D], dtype=pl.FP32)
            T1 = pl.create_tensor([P, D, D], dtype=pl.FP32)
            T2 = pl.create_tensor([P, D, D], dtype=pl.FP32)

            for r in pl.range(P):
                fdst = pld.window(fdst_buf, [D, D], dtype=pl.FP32)
                fsig = pld.window(fsig_buf, [S, 1], dtype=pl.INT32)
                bdst = pld.window(bdst_buf, [D, D], dtype=pl.FP32)
                bsig = pld.window(bsig_buf, [S, 1], dtype=pl.INT32)

                if r == 0:
                    a0 = self.c_compute(X[r], T0[r], device=r)
                    self.c_send(a0, Echo[r], fdst, fsig, r + 1, device=r)   # -> rank 1
                    a1 = self.c_compute(a0, T1[r], device=r)
                    self.c_recv(Rb[r], bdst, bsig, device=r)                # <- rank 1
                    self.c_compute(a1, T2[r], device=r)
                else:
                    b0 = self.c_compute(X[r], T0[r], device=r)
                    self.c_send(b0, Echo[r], bdst, bsig, r - 1, device=r)   # -> rank 0
                    b1 = self.c_compute(b0, T1[r], device=r)
                    self.c_recv(Rf[r], fdst, fsig, device=r)                # <- rank 0
                    self.c_compute(b1, T2[r], device=r)
            return Echo, Rf, Rb

    return TwoRing


def main() -> int:
    devices = [int(x) for x in (sys.argv[1] if len(sys.argv) > 1 else "0,1").split(",")][:P]
    torch.manual_seed(5)
    X = torch.randn(P, D, D).share_memory_()
    Echo = torch.zeros(P, D, D).share_memory_()
    Rf = torch.zeros(P, D, D).share_memory_()
    Rb = torch.zeros(P, D, D).share_memory_()

    compiled = ir.compile(build(), platform="a2a3",
                          distributed_config=DistributedConfig(device_ids=devices,
                                                               num_sub_workers=0))
    rt = compiled.prepare()
    try:
        t0 = time.time()
        rt(X, Echo, Rf, Rb)
        fwd_ok = (Rf[1] - X[0] * 2.0).abs().max().item() < 1e-6
        rev_ok = (Rb[0] - X[1] * 2.0).abs().max().item() < 1e-6
        print(f"COMPLETED in {time.time() - t0:.1f}s  fwd_ok={fwd_ok} rev_ok={rev_ok}")
    finally:
        rt.close()
    return 0


if __name__ == "__main__":
    sys.exit(main())
```

</details>

### Expected Behavior

Dispatches issued in sequence from `host_orch` for a given rank should execute in that order — or, if that is explicitly not guaranteed for comm dispatches, the program should be rejected at compile time rather than hanging. `MaterializeCommDomainScopes` already rejects other comm-structure mistakes with clear messages (`materialize_comm_domain_scopes_pass.cpp:249` "device=r over a non-unit-step loop is not supported", and `:261` "device= Var is not the induction variable of any enclosing pl.range loop"), so this class of error has precedent for a compile-time diagnostic.

### Actual Behavior

Both ranks stall and the run fails with:

```
[ERROR] PTO2 scheduler timeout sub_class=S1:running-stalled (detail=1) \
        completed=0/1 running=1 ready=0 waiting=0 orch_done=1
RuntimeError: ... finalize_native_run failed with code -100 \
        (run_id=1 slot=0 generation=1 dispatch_id=2 run_epoch=2)
```

Two details pin the cause:

- `waiting=0` — nothing was blocked on fan-in. The stalled task had no unmet dependency; it was simply dispatched too early.
- `dispatch_id=2` — the stall is at the **second** dispatch on the rank. A healthy run reports `dispatch_id=5`. After the first compute, the next thing dispatched is the wait, not the send.

### Additional Context

#### Why this happens, in the runtime's own terms

- `runtime/docs/orchestrator.md` §7: dependencies are RAW/WAW over a producer-keyed tensormap, driven by `TaskArgs` tags. Nothing else creates an edge.
- `runtime/docs/scheduler.md`: *"a task is routed only once all producers complete"*, and *"both immediately-ready submissions and dependency-released consumers use the same routing operation"* — so the FIFO is ordered by **readiness, not submission**.
- `runtime/docs/directed-next-level-scheduling.md`: *"Submission records each live producer in the consumer's `fanin_count`"* (zero producers ⇒ READY at submission), and dispatch happens only when the target worker is idle, strict FIFO head, with *"no idle-worker search, rebinding, work stealing, or scan into another worker's queue"*.

The same page states the contract this violates: *"A NEXT_LEVEL single does not wait for a not-yet-dispatched peer at the same scheduler level. Work that requires concurrent placement uses the group API."* The distributed codegen emits only singles (`_submit_chip` → `submit_next_level`), never `submit_next_level_group`.

In the generated `host_orch.py`, rank 0's `c_recv` receives only its `Out` tensor plus the two windows — no argument that any earlier dispatch produces — so its `fanin_count` is 0 and it enters `FIFO[0]` immediately, behind only the first compute and **ahead of `c_send`**, which is still pending on that compute.

#### It is not the explanations one reaches for first

Full factorial at P=2, each cell measured, controls holding buffer count and dispatch count fixed:

| arm | buffers | dispatches | every rank waits? | send+wait in one dispatch? | result |
| --- | --- | --- | --- | --- | --- |
| one ring | 2 | 4 | no | – | passes |
| one ring, padded to 5 dispatches | 2 | 5 | no | – | passes |
| two rings, SAME direction | 4 | 5 | no | – | passes |
| two rings, comm fused per rank | 4 | 5 | **yes** | **yes** | **passes** |
| two rings, comm split (`repro.py`) | 4 | 5 | **yes** | **no** | **deadlock** |

So it is not "two comm windows in one program", not the buffer count, not the dispatch count, and not the cyclic wait as such — it is precisely *a rank's wait sitting in a different dispatch from its send*.

It is also not the notify/wait ordering rule in `docs/en/user/distributed/04-debugging.md` (*"ensure every rank calls notify before any rank calls wait"*): the reproducer **obeys** that rule and deadlocks, while the fused-comm arm **violates** it (wait-then-send inside one kernel) and passes, as does an AllScan ring middle step with the same wait-then-send shape that is HW-validated at P=4.

#### Direct confirmation

Adding **one** artificial RAW edge — the send's `Out` tensor threaded into the recv as an extra `INPUT`, made payload-neutral by storing `dst + (dep - dep)` — with nothing else changed turns the deadlock into a **2.3 s pass, 4/4 runs**, while the unmodified controls stall **2/2**. That is the missing fan-in, isolated with no other variable moved.

#### Impact

Any multi-phase distributed program where a rank's comm phases are split across dispatches. It blocked a fully-fused sequence-parallel attention-style backward (recompute → forward ring → gradient → reverse ring → gradient) at *every* `P>1` configuration, while the same operator was correct at `P=1`. Programs that keep all of a rank's comm inside a single `chip_orch` dispatch — which is what an AllScan ring and `tests/st/distributed/test_l3_ep_dispatch_combine.py` do — are unaffected, which is why this has not been hit before.

A fix is proposed in a companion PR.

--- a/python/pypto/runtime/distributed_runner.py
+++ b/python/pypto/runtime/distributed_runner.py
@@ -448,6 +448,31 @@ def _load_orch_entry(output_dir: Path) -> tuple[Any, Any]:
     return entry_fn, alloc_fn
 
 
+def _call_alloc_intermediates(alloc_fn: Any, tensors: dict[str, Any], world_size: int) -> None:
+    """Invoke the generated ``_alloc_intermediates``, passing ``world_size`` only if it
+    accepts it.
+
+    Codegen emits ``_alloc_intermediates(tensors, world_size=1)`` so it can size the
+    per-rank comm ordering tokens, but the one-argument form must keep working: a
+    ``build_output/`` directory produced by an older pypto is replayable via
+    :meth:`DistributedCompiledProgram.from_dir`, and callers may inject their own
+    allocator. Probing the signature (rather than catching :class:`TypeError` around the
+    call) keeps a ``TypeError`` raised *inside* the allocator from being mistaken for a
+    signature mismatch and silently retried.
+    """
+    try:
+        params = inspect.signature(alloc_fn).parameters
+    except (TypeError, ValueError):  # builtin / C callable with no introspectable signature
+        params = {}
+    takes_world_size = "world_size" in params or any(
+        p.kind is inspect.Parameter.VAR_KEYWORD for p in params.values()
+    )
+    if takes_world_size:
+        alloc_fn(tensors, world_size=world_size)
+    else:
+        alloc_fn(tensors)
+
+
 def _load_sub_worker_fns(output_dir: Path) -> dict[str, Any]:
     """Load SubWorker callables from ``sub_workers/*.py`` (keyed by file stem)."""
     sub_worker_fns: dict[str, Any] = {}
@@ -1291,7 +1316,7 @@ def execute_distributed(
     # ``world_size`` sizes the per-rank comm ordering tokens, which must exist
     # pre-fork like every other host-side intermediate.
     if alloc_fn is not None:
-        alloc_fn(tensors, world_size=len(dc.device_ids))
+        _call_alloc_intermediates(alloc_fn, tensors, len(dc.device_ids))
 
     sub_worker_fns = _load_sub_worker_fns(output_dir)
     # The one-shot path cannot supply callbacks; if the program declares any
@@ -1580,7 +1605,9 @@ class DistributedWorker(Worker):
                 for _frame in self._dispatch_frames:
                     base_tensors: dict[str, Any] = {}
                     if alloc_fn is not None:
-                        alloc_fn(base_tensors, world_size=len(prog._distributed_config.device_ids))
+                        _call_alloc_intermediates(
+                            alloc_fn, base_tensors, len(prog._distributed_config.device_ids)
+                        )
                     base_tensor_frames.append(base_tensors)
                 self._states[prog] = {
                     "entry_fn": entry_fn,

--- a/python/pypto/runtime/distributed_runner.py
+++ b/python/pypto/runtime/distributed_runner.py
@@ -1288,8 +1288,10 @@ def execute_distributed(
 
     # Pre-fork: allocate HOST-level intermediate tensors so the POSIX
     # shared-memory mappings exist before w.init() forks child processes.
+    # ``world_size`` sizes the per-rank comm ordering tokens, which must exist
+    # pre-fork like every other host-side intermediate.
     if alloc_fn is not None:
-        alloc_fn(tensors)
+        alloc_fn(tensors, world_size=len(dc.device_ids))
 
     sub_worker_fns = _load_sub_worker_fns(output_dir)
     # The one-shot path cannot supply callbacks; if the program declares any
@@ -1578,7 +1580,7 @@ class DistributedWorker(Worker):
                 for _frame in self._dispatch_frames:
                     base_tensors: dict[str, Any] = {}
                     if alloc_fn is not None:
-                        alloc_fn(base_tensors)
+                        alloc_fn(base_tensors, world_size=len(prog._distributed_config.device_ids))
                     base_tensor_frames.append(base_tensors)
                 self._states[prog] = {
                     "entry_fn": entry_fn,

--- a/src/codegen/distributed/distributed_codegen.cpp
+++ b/src/codegen/distributed/distributed_codegen.cpp
@@ -1043,10 +1043,16 @@ void DistributedCodegen::EmitCallToWorker(const ir::CallPtr& call, const ir::Fun
   // Empty string means no ``device=`` was set — comm-less L3 dispatch path.
   std::string rank_expr = ResolveRankExpr(call);
 
-  // Set when any arg is a DistributedTensor, i.e. this dispatch touches a comm
-  // window and therefore participates in the per-rank comm ordering chain
-  // emitted by VisitStmt_(CommDomainScopeStmtPtr).
-  bool touches_comm_window = false;
+  // Every comm domain whose window buffers this dispatch references, in first-reference
+  // order and deduplicated. One ordering token is appended per domain below.
+  //
+  // Keyed on each buffer's OWNING scope (``ScopeForWindowBuffer``), not on the innermost
+  // lexical scope: a dispatch nested inside an inner domain may consume an outer domain's
+  // buffer, and this path — unlike the builtin-collective emitter, which asserts all its
+  // window args share one scope — places no such restriction on user kernels. Attaching the
+  // WAW edge to the lexically-innermost scope would then chain the dispatch into a domain it
+  // does not actually touch while leaving the domain it *does* touch unordered.
+  std::vector<ir::CommDomainScopeStmtPtr> comm_arg_scopes;
 
   // Build TaskArgs from callee's parameter directions
   emitter_.EmitLine(ta_var + " = TaskArgs()");
@@ -1087,8 +1093,13 @@ void DistributedCodegen::EmitCallToWorker(const ir::CallPtr& call, const ir::Fun
       if (i < callee->param_directions_.size()) {
         tag = ParamDirectionToTensorArgType(callee->param_directions_[i]);
       }
-      touches_comm_window = true;
-      const std::string handle_var = HandleVarForScope(ScopeForWindowBuffer(window_buffer));
+      const auto owning_scope = ScopeForWindowBuffer(window_buffer);
+      if (std::none_of(
+              comm_arg_scopes.begin(), comm_arg_scopes.end(),
+              [&](const ir::CommDomainScopeStmtPtr& seen) { return seen.get() == owning_scope.get(); })) {
+        comm_arg_scopes.push_back(owning_scope);
+      }
+      const std::string handle_var = HandleVarForScope(owning_scope);
       emitter_.EmitLine(ta_var + ".add_tensor(" + handle_var + "[" + rank_expr + "].buffers[\"" + name +
                         "\"].tensor(shapes=" + shape + ", dtype=" + dtype_enum + "), " + tag + ")");
       continue;
@@ -1167,10 +1178,12 @@ void DistributedCodegen::EmitCallToWorker(const ir::CallPtr& call, const ir::Fun
   // expected_arg_count`, so an extra arg is accepted and the generated
   // orchestration entry simply does not index it. That is what keeps this fix
   // host-side only — no chip .cpp regeneration, no `expected_arg_count` change.
-  if (!is_sub && touches_comm_window && !rank_expr.empty() && !comm_domain_stack_.empty()) {
-    const std::string ord_key = HandleVarForScope(comm_domain_stack_.back()) + "_ord";
-    emitter_.EmitLine(ta_var + ".add_tensor(make_tensor_arg(tensors[\"" + ord_key + "\"][" + rank_expr +
-                      ", 0:1]), TensorArgType.INOUT)");
+  if (!is_sub && !rank_expr.empty()) {
+    for (const auto& scope : comm_arg_scopes) {
+      const std::string ord_key = HandleVarForScope(scope) + "_ord";
+      emitter_.EmitLine(ta_var + ".add_tensor(make_tensor_arg(orch._worker, tensors[\"" + ord_key + "\"][" +
+                        rank_expr + "]), TensorArgType.INOUT)");
+    }
   }
 
   for (const auto& line : scalar_lines) {
@@ -1436,11 +1449,20 @@ void DistributedCodegen::EmitAllocIntermediatesFunction(const ir::FunctionPtr& h
   // comm dispatches a WAW chain in program order. It costs no parallelism: a
   // worker executes one task at a time regardless, so it constrains ORDER, not
   // concurrency.
+  //
+  // One SEPARATE allocation per rank, held in a list -- not rows of a single tensor. The
+  // runtime memoizes a host-tensor handle by its storage base and keys dependencies on that
+  // identity, so every view of one storage collapses into a single dependency node
+  // (``Worker.make_tensor_arg``: "every ref over the same storage shares one canonical
+  // identity and dependencies key on it"). Slicing one tensor per rank would therefore chain
+  // ALL ranks into one order, not each rank into its own -- and a program whose ranks must be
+  // in flight together, such as a barrier, cannot make progress under that.
   std::vector<ir::CommDomainScopeStmtPtr> comm_scopes;
   if (host_orch) CollectCommDomainScopes(host_orch->body_, &comm_scopes);
   for (const auto& scope : comm_scopes) {
     emitter_.EmitLine("tensors[\"" + HandleVarForScope(scope) +
-                      "_ord\"] = torch.zeros((max(world_size, 1), 1), dtype=torch.int32).share_memory_()");
+                      "_ord\"] = [torch.zeros((1,), dtype=torch.int32).share_memory_() "
+                      "for _ in range(max(world_size, 1))]");
   }
 
   if (hoisted_allocs_.empty() && comm_scopes.empty()) {

--- a/src/codegen/distributed/distributed_codegen.cpp
+++ b/src/codegen/distributed/distributed_codegen.cpp
@@ -605,6 +605,12 @@ void DistributedCodegen::VisitStmt_(const ir::CommDomainScopeStmtPtr& op) {
   emitter_.EmitLine(") as " + handle_var + ":");
   emitter_.IncreaseIndent();
 
+  // The per-rank comm ordering token for this scope is allocated in
+  // ``_alloc_intermediates`` (see EmitAllocIntermediates), NOT here: it must
+  // exist as a POSIX shared mapping BEFORE ``w.init()`` forks the chip
+  // children, or the children cannot see it and staging the arg fails with
+  // "Failed to stage tensor N to device".
+
   // Push this scope so EmitCallToWorker can resolve DistributedTensor args
   // inside the body to the correct handle var. Pop on exit so a sibling
   // scope's body never sees this one.
@@ -1037,13 +1043,25 @@ void DistributedCodegen::EmitCallToWorker(const ir::CallPtr& call, const ir::Fun
   // Empty string means no ``device=`` was set — comm-less L3 dispatch path.
   std::string rank_expr = ResolveRankExpr(call);
 
+  // Set when any arg is a DistributedTensor, i.e. this dispatch touches a comm
+  // window and therefore participates in the per-rank comm ordering chain
+  // emitted by VisitStmt_(CommDomainScopeStmtPtr).
+  bool touches_comm_window = false;
+
   // Build TaskArgs from callee's parameter directions
   emitter_.EmitLine(ta_var + " = TaskArgs()");
 
+  // ``TaskArgs`` requires every tensor to precede every scalar ("cannot add
+  // tensor after scalar"), so scalar emission is buffered and flushed once the
+  // tensor list is complete — otherwise the trailing comm ordering token below
+  // would be rejected at run time. Order within the scalar list is preserved,
+  // so the layout still matches the callee parameter list.
+  std::vector<std::string> scalar_lines;
+
   for (size_t i = 0; i < call->args_.size(); ++i) {
     if (ir::IsA<ir::CommCtxType>(call->args_[i]->GetType())) {
-      emitter_.EmitLine(ta_var + ".add_scalar(" + ResolveCommCtxArg(call->args_[i], rank_expr, call->span_) +
-                        ")");
+      scalar_lines.push_back(ta_var + ".add_scalar(" +
+                             ResolveCommCtxArg(call->args_[i], rank_expr, call->span_) + ")");
       continue;
     }
 
@@ -1069,6 +1087,7 @@ void DistributedCodegen::EmitCallToWorker(const ir::CallPtr& call, const ir::Fun
       if (i < callee->param_directions_.size()) {
         tag = ParamDirectionToTensorArgType(callee->param_directions_[i]);
       }
+      touches_comm_window = true;
       const std::string handle_var = HandleVarForScope(ScopeForWindowBuffer(window_buffer));
       emitter_.EmitLine(ta_var + ".add_tensor(" + handle_var + "[" + rank_expr + "].buffers[\"" + name +
                         "\"].tensor(shapes=" + shape + ", dtype=" + dtype_enum + "), " + tag + ")");
@@ -1089,7 +1108,7 @@ void DistributedCodegen::EmitCallToWorker(const ir::CallPtr& call, const ir::Fun
     // ScalarType formal: pass through via ``add_scalar`` in IR-argument order
     // so the runtime TaskArgs layout matches the callee parameter list.
     if (ir::As<ir::ScalarType>(call->args_[i]->GetType())) {
-      emitter_.EmitLine(ta_var + ".add_scalar(" + arg_str + ")");
+      scalar_lines.push_back(ta_var + ".add_scalar(" + arg_str + ")");
       continue;
     }
 
@@ -1138,6 +1157,24 @@ void DistributedCodegen::EmitCallToWorker(const ir::CallPtr& call, const ir::Fun
       emitter_.EmitLine(ta_var + ".add_tensor(make_tensor_arg(orch._worker, tensors[\"" + target +
                         "\"]), TensorArgType.OUTPUT_EXISTING)");
     }
+  }
+
+  // Append the per-rank comm ordering token as the LAST TENSOR — after every
+  // real tensor (so their indices are unchanged) but before the scalars, which
+  // TaskArgs requires. Tensor and scalar args are indexed independently, so a
+  // trailing tensor shifts nothing. The chip side never reads it:
+  // `aicpu_orchestration_config` validates `actual_arg_count <
+  // expected_arg_count`, so an extra arg is accepted and the generated
+  // orchestration entry simply does not index it. That is what keeps this fix
+  // host-side only — no chip .cpp regeneration, no `expected_arg_count` change.
+  if (!is_sub && touches_comm_window && !rank_expr.empty() && !comm_domain_stack_.empty()) {
+    const std::string ord_key = HandleVarForScope(comm_domain_stack_.back()) + "_ord";
+    emitter_.EmitLine(ta_var + ".add_tensor(make_tensor_arg(tensors[\"" + ord_key + "\"][" + rank_expr +
+                      ", 0:1]), TensorArgType.INOUT)");
+  }
+
+  for (const auto& line : scalar_lines) {
+    emitter_.EmitLine(line);
   }
 
   if (is_sub) {
@@ -1304,6 +1341,34 @@ void RejectNestedTensorCreate(const ir::StmtPtr& stmt, const std::string& contai
   }
 }
 
+// Collects every CommDomainScopeStmt reachable in a body, in source order.
+// Used to allocate each scope's per-rank comm ordering token inside
+// ``_alloc_intermediates`` — those allocations must happen pre-fork, so they
+// cannot be emitted from VisitStmt_(CommDomainScopeStmtPtr), which runs while
+// emitting host_orch's body.
+void CollectCommDomainScopes(const ir::StmtPtr& stmt, std::vector<ir::CommDomainScopeStmtPtr>* out) {
+  if (!stmt) return;
+  if (auto scope = std::dynamic_pointer_cast<const ir::CommDomainScopeStmt>(stmt)) {
+    out->push_back(scope);
+    CollectCommDomainScopes(scope->body_, out);
+    return;
+  }
+  if (auto seq = std::dynamic_pointer_cast<const ir::SeqStmts>(stmt)) {
+    for (const auto& s : seq->stmts_) CollectCommDomainScopes(s, out);
+    return;
+  }
+  if (auto for_stmt = std::dynamic_pointer_cast<const ir::ForStmt>(stmt)) {
+    CollectCommDomainScopes(for_stmt->body_, out);
+    return;
+  }
+  if (auto if_stmt = std::dynamic_pointer_cast<const ir::IfStmt>(stmt)) {
+    CollectCommDomainScopes(if_stmt->then_body_, out);
+    if (if_stmt->else_body_.has_value()) {
+      CollectCommDomainScopes(*if_stmt->else_body_, out);  // NOLINT(bugprone-unchecked-optional-access)
+    }
+  }
+}
+
 // Returns the immediate top-level statements of a function body. The HOST
 // orchestrator body is a SeqStmts in practice; the single-stmt case is
 // handled for safety.
@@ -1347,11 +1412,40 @@ void DistributedCodegen::CollectHostOrchHoistableAllocs(const ir::FunctionPtr& h
 }
 
 void DistributedCodegen::EmitAllocIntermediatesFunction(const ir::FunctionPtr& host_orch) {
-  emitter_.EmitLine("def _alloc_intermediates(tensors):");
+  // ``world_size`` is passed by the runtime (distributed_runner) and defaults
+  // to 1 so an older caller still loads. It sizes the per-rank comm ordering
+  // tokens below, which cannot be sized at codegen time when a domain's worker
+  // list is `*range(world_size)`.
+  emitter_.EmitLine("def _alloc_intermediates(tensors, world_size=1):");
   emitter_.IncreaseIndent();
-  if (hoisted_allocs_.empty()) {
+
+  // Per-rank comm ordering tokens, one per comm domain scope.
+  //
+  // A chip dispatch becomes one `submit_next_level` DAG node whose edges come
+  // ONLY from tensor tags. A comm window carries no such edge, so dispatches
+  // that interact solely through `remote_store`/`notify`/`wait` are independent
+  // to the scheduler and the program order written in host_orch is discarded.
+  // The runtime routes a task to its per-worker FIFO as soon as it is READY
+  // (runtime/docs/scheduler.md: immediately-ready submissions and
+  // dependency-released consumers share one router), and a dispatch whose only
+  // job is to wait usually has NO producer — so it is routed ahead of the same
+  // rank's still-pending send. One task per worker means that spin-wait then
+  // owns the core and the send never runs: deadlock.
+  //
+  // Threading this token through every comm dispatch as INOUT gives a rank's
+  // comm dispatches a WAW chain in program order. It costs no parallelism: a
+  // worker executes one task at a time regardless, so it constrains ORDER, not
+  // concurrency.
+  std::vector<ir::CommDomainScopeStmtPtr> comm_scopes;
+  if (host_orch) CollectCommDomainScopes(host_orch->body_, &comm_scopes);
+  for (const auto& scope : comm_scopes) {
+    emitter_.EmitLine("tensors[\"" + HandleVarForScope(scope) +
+                      "_ord\"] = torch.zeros((max(world_size, 1), 1), dtype=torch.int32).share_memory_()");
+  }
+
+  if (hoisted_allocs_.empty() && comm_scopes.empty()) {
     emitter_.EmitLine("pass");
-  } else {
+  } else if (!hoisted_allocs_.empty()) {
     // Walk top-level statements in original order; emit only hoisted allocs.
     // Falls through the normal AssignStmt → Call → EmitTensorCreate path.
     // host_orch_body_after_hoist_ is FALSE here so EmitTensorCreate runs

--- a/src/codegen/distributed/distributed_ops_codegen.cpp
+++ b/src/codegen/distributed/distributed_ops_codegen.cpp
@@ -182,6 +182,16 @@ void EmitBuiltinWindowCollectiveDispatch(DistributedCodegen& codegen, const Call
         << "Internal error: unsupported builtin tensor collective arg type at index " << i;
   }
 
+  // Per-rank comm ordering token, appended as the LAST TENSOR (before the scalars, which
+  // TaskArgs requires). Builtin collectives submit through their own submit_next_level, so
+  // without this they sit outside the ordering chain that EmitCallToWorker builds: a rank
+  // mixing a builtin barrier/collective with a custom comm kernel — or issuing builtins
+  // across separate dispatches — would still let a waiting dispatch be routed ahead of its
+  // producer, which is the deadlock this token exists to prevent. All window args here are
+  // checked above to share one comm domain, so a single token is enough.
+  codegen.Emit(ta_var + ".add_tensor(make_tensor_arg(orch._worker, tensors[\"" + *handle_var + "_ord\"][" +
+               rank_expr + "]), TensorArgType.INOUT)");
+
   codegen.Emit(ta_var + ".add_scalar(" + *handle_var + "[" + rank_expr + "].domain_size)");
   codegen.Emit(ta_var + ".add_scalar(" + *handle_var + "[" + rank_expr + "].device_ctx)");
   if (core_num.has_value()) {

--- a/tests/st/distributed/test_l3_host_tensor_allgather.py
+++ b/tests/st/distributed/test_l3_host_tensor_allgather.py
@@ -33,6 +33,7 @@ cross-process data race.
 ST coverage: P=2 and P=4 (skips when fewer devices are available).
 """
 
+import re
 import sys
 
 import pypto.language as pl
@@ -248,6 +249,42 @@ class TestL3HostTensorAllGather:
         variant_dir = compiled.output_dir / "next_levels" / "builtin.tensor.allgather__fp32"
         assert variant_dir.is_dir()
         assert (variant_dir / "kernel_config.py").is_file()
+
+        # The builtin collective must join the per-rank comm ordering chain. It submits via
+        # its own orch.submit_next_level in EmitBuiltinWindowCollectiveDispatch rather than
+        # the custom-kernel path, so without the ordering token it would sit outside the
+        # chain -- and a rank mixing a builtin collective with a custom comm kernel (exactly
+        # this program: publish_orch -> allgather -> consume_orch) could then have a waiting
+        # dispatch routed ahead of its producer and deadlock.
+        orch_src = (compiled.output_dir / "orchestration" / "host_orch.py").read_text()
+        # One separate allocation per rank, not rows of a single tensor. The runtime keys a
+        # host tensor's dependencies on its storage base, so every view of one storage fuses
+        # into a single node -- rows of one tensor would collapse the per-rank chains into one
+        # global order, which a program whose ranks must be in flight together cannot satisfy.
+        assert re.search(r'_ord"\] = \[torch\.zeros\(.*for _ in range\(', orch_src), (
+            "comm ordering token is not a per-rank list of separate allocations"
+        )
+        submits = [
+            line
+            for line in orch_src.splitlines()
+            if "_submit_chip(" in line or "orch.submit_next_level(" in line
+        ]
+        assert any("builtin.tensor.allgather" in line for line in submits), submits
+        for submit in submits:
+            ta_var = re.search(r"(_ta_\d+)", submit).group(1)
+            token_lines = [
+                line.strip()
+                for line in orch_src.splitlines()
+                if line.strip().startswith(f"{ta_var}.add_tensor(") and '_ord"][' in line
+            ]
+            assert token_lines, f"dispatch without an ordering token: {submit.strip()}"
+            # The token goes through the same make_tensor_arg(worker, tensor) form as every
+            # other host tensor arg. Asserting the form, not just the presence, is what
+            # separates "no token" from "token emitted as an uncallable line".
+            for token in token_lines:
+                assert "make_tensor_arg(orch._worker, " in token, (
+                    f"ordering token is not a well-formed make_tensor_arg call: {token}"
+                )
 
         inputs = _make_rank_inputs(n_ranks)
         outputs = torch.zeros((n_ranks, 1, n_ranks, SIZE), dtype=torch.float32)

--- a/tests/ut/codegen/test_distributed_codegen.py
+++ b/tests/ut/codegen/test_distributed_codegen.py
@@ -551,7 +551,7 @@ class TestDistributedCodegen:
         cg = codegen.DistributedCodegen()
         code = cg.generate(program)
 
-        alloc_idx = code.find("def _alloc_intermediates(tensors):")
+        alloc_idx = code.find("def _alloc_intermediates(tensors, world_size=1):")
         host_idx = code.find("def host_orch(")
         assert alloc_idx >= 0, f"Missing _alloc_intermediates in:\n{code}"
         assert host_idx >= 0, f"Missing host_orch in:\n{code}"
@@ -600,9 +600,9 @@ class TestDistributedCodegen:
         cg = codegen.DistributedCodegen()
         code = cg.generate(program)
 
-        assert "def _alloc_intermediates(tensors):" in code
+        assert "def _alloc_intermediates(tensors, world_size=1):" in code
         # Body is just `pass` since there are no allocations to hoist.
-        alloc_idx = code.find("def _alloc_intermediates(tensors):")
+        alloc_idx = code.find("def _alloc_intermediates(tensors, world_size=1):")
         host_idx = code.find("def host_orch(")
         alloc_block = code[alloc_idx:host_idx]
         assert "    pass" in alloc_block

--- a/tests/ut/runtime/test_distributed_worker.py
+++ b/tests/ut/runtime/test_distributed_worker.py
@@ -1664,6 +1664,38 @@ class TestLoadOrchEntry:
         assert entry_fn.__name__ == "host_orch"
         assert alloc_fn is not None and alloc_fn.__name__ == "_alloc_intermediates"
 
+    def test_alloc_intermediates_accepts_both_signatures(self, tmp_path):
+        """The runtime must call a one-arg allocator without ``world_size``.
+
+        Codegen emits ``_alloc_intermediates(tensors, world_size=1)`` so it can size the
+        per-rank comm ordering tokens, but a ``build_output/`` produced by an older pypto
+        is still replayable via ``from_dir``, and callers may inject their own allocator.
+        Both shapes must work.
+        """
+        from pypto.runtime.distributed_runner import (  # noqa: PLC0415
+            _call_alloc_intermediates,
+        )
+
+        seen: dict[str, object] = {}
+
+        def old_style(tensors):
+            seen["old"] = True
+
+        def new_style(tensors, world_size=1):
+            seen["new"] = world_size
+
+        _call_alloc_intermediates(old_style, {}, 4)
+        _call_alloc_intermediates(new_style, {}, 4)
+        assert seen == {"old": True, "new": 4}
+
+        # A TypeError raised *inside* the allocator must propagate, not be mistaken for a
+        # signature mismatch and retried with fewer arguments.
+        def boom(tensors, world_size=1):
+            raise TypeError("inner failure")
+
+        with pytest.raises(TypeError, match="inner failure"):
+            _call_alloc_intermediates(boom, {}, 4)
+
     def test_no_marker_raises(self, tmp_path):
         from pypto.runtime.distributed_runner import _load_orch_entry  # noqa: PLC0415
 

